diff --git a/formats/lttng-live/Makefile.am b/formats/lttng-live/Makefile.am
index c834699..0142e1f 100644
--- a/formats/lttng-live/Makefile.am
+++ b/formats/lttng-live/Makefile.am
@@ -1,4 +1,4 @@
-AM_CFLAGS = $(PACKAGE_CFLAGS) -I$(top_srcdir)/include -I$(top_builddir)/include
+AM_CFLAGS = $(PACKAGE_CFLAGS) -I$(top_srcdir)/include -I$(top_srcdir) -I$(top_builddir)/include
 
 lib_LTLIBRARIES = libbabeltrace-lttng-live.la
 
