diff --git a/urcu/compiler.h b/urcu/compiler.h
index 1e30903..f38e49c 100644
--- a/urcu/compiler.h
+++ b/urcu/compiler.h
@@ -122,7 +122,19 @@
  */
 # ifdef __ARMEL__
 #  if URCU_GCC_VERSION >= 40800 && URCU_GCC_VERSION <= 40802
-#   error Your gcc version produces clobbered frame accesses
+#   ifdef __CS_SOURCERYGXX_MAJ__
+#    if __CS_SOURCERYGXX_MAJ__ < 2013
+#     error Sourcery CodeBench 2013.11-32 or higher is required
+#    endif
+#    if __CS_SOURCERYGXX_MAJ__ == 2013 && __CS_SOURCERYGXX_MIN__ < 11
+#     error Sourcery CodeBench 2013.11-32 or higher is required
+#    endif
+#    if __CS_SOURCERYGXX_MAJ__ == 2013 && __CS_SOURCERYGXX_MIN__ == 11 && __CS_SOURCERYGXX_REV__ < 32
+#     error Sourcery CodeBench 2013.11-32 or higher is required
+#    endif
+#   else
+#    error Your gcc version produces clobbered frame accesses
+#   endif
 #  endif
 # endif
 #endif
